src/api/resources/sync/views.py

```diff
@@ -0,0 +1,348 @@
+
+    return vw_request
+
+
+def _get_source(source_type):
+    return project_sources.get(source_type) or non_project_sources.get(source_type)
+
+
+def _get_required_parameters(request, data, required_filters):
+    params = {"revision_num": data.get("last_revision")}
+
+    if "project" in required_filters:
+        project = data.get("project")
+        if project is None:
+            raise ValueError("Project is missing")
+        params["project"] = project
+
+    if "profile" in required_filters:
+        user = request.user
+        if user is None:
+            raise ValueError("Profile is missing")
+
+        params["profile"] = str(user.profile.pk)
+
+    return params
+
+
+def _validate_source_types(data):
+    return [st for st in data if _get_source(st) is None]
+
+
+def _get_choices():
+    src = _get_source(CHOICES_SOURCE_TYPE)
+    return {
+        "updates": src["view"]().get_choices(),
+        "deletes": [],
+        "last_revision_num": -1,
+    }
+
+
+def _error(status_code, exception, data=None):
+    return {"status_code": status_code, "message": str(exception), "data": data}
+
+
+def _format_errors(errors):
+    if errors is None:
+        return None
+
+    return {k: v[0] for k, v in errors.items()}
+
+
+def _get_serialized_record(serializer, record_id):
+    record = get_record(model_class=serializer.Meta.model, record_id=record_id)
+
+    data = serialize_revisions(serializer, [record])
+    if data["updates"]:
+        return data["updates"][0]
+    elif data["deletes"]:
+        return data["deletes"][0]
+
+    return None
+
+
+def _update_source_record(source_type, serializer, record, request):
+    src = _get_source(source_type)
+    vw_request = _create_view_request(
+        request, method=get_request_method(record), data=record
+    )
+
+    record_id = record.get("id")
+    failed_permission_checks = check_permissions(
+        vw_request, {source_type: record}, [source_type]
+    )
+
+    if failed_permission_checks:
+        status_code, _ = failed_permission_checks[0]
+        exception = NotAuthenticated if status_code == 401 else PermissionError
+        return _error(403, exception())
+
+    if record_id is None:
+        return _error(
+            400, ValidationError(), data={"id": "This field may not be null."}
+        )
+
+    if src["read_only"] is True:
+        return _error(405, ReadOnlyError(f"{source_type} is read-only"))
+
+    try:
+        status_code, errors = apply_changes(vw_request, serializer, record)
+        if status_code == 400:
+            msg = "Validation Error"
+            data = _format_errors(errors)
+        elif status_code == 409:
+            msg = "Conflict"
+            data = _get_serialized_record(serializer, record_id)
+        else:
+            msg = ""
+            data = _get_serialized_record(serializer, record_id)
+
+        return {"status_code": status_code, "message": msg, "data": data}
+    except Exception as err:
+        return _error(500, err)
```

Comment:
Is there anywhere else we should be returning a 500? If it's really on the level of a 500, should we attempt to hook into logging so we're notified the same way?